File: euler/p11.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"strconv"
	"strings"
)

const dim = 20
const chunk = 4

var matrix [][]uint64

func max( nums ... uint64) uint64 {
	var max_value uint64 = 0
	for _, num := range nums {
		if num > max_value {
			fmt.Printf("%d\n", num)
			max_value = num
		}
	}	
	return max_value
}

func left(x int, y int) uint64 {
	var prod uint64 = 0
	boundx := chunk -1
	if x >= boundx {
		prod=1
		for i:=0; i<chunk; i++ {
			prod = prod*matrix[boundx-i][y]
		}
	}
	return prod
}

func right(x int, y int) uint64 {
	var prod uint64 = 0
	boundx := dim - chunk
	if x <= boundx {
		prod=1
		for i:=0; i<chunk; i++ {
			prod = prod*matrix[boundx+i][y]
		}
	}
	return prod
}

func up(x int, y int) uint64 {
	var prod uint64 = 0
	boundy := chunk -1
	if y >= boundy {
		prod=1
		for i:=0; i<chunk; i++ {
			prod = prod*matrix[x][boundy-i]
		}
	}
	return prod
}

func down(x int, y int) uint64 {
	var prod uint64 = 0
	boundy := dim - chunk
	if y <= boundy {
		prod=1
		for i:=0; i<chunk; i++ {
			prod = prod*matrix[x][boundy+i]
		}
	}
	return prod
}

func up_left(x int, y int) uint64 {
	var prod uint64 = 0
	boundx := chunk - 1
	boundy := chunk - 1
	if y >= boundy && x>=boundx{
		prod=1
		for i:=0; i<chunk; i++ {
			prod = prod*matrix[boundx-i][boundy-i]
		}
	}
	return prod
}

func up_right(x int, y int) uint64 {
	var prod uint64 = 0
	boundx := dim - chunk
	boundy := chunk - 1
	if y >= boundy && x<=boundx{
		prod=1
		for i:=0; i<chunk; i++ {
			prod = prod*matrix[boundx+i][boundy-i]
		}
	}
	return prod
}

func down_left(x int, y int) uint64 {
	var prod uint64 = 0
	boundx := chunk -1
	boundy := dim - chunk
	if y <= boundy && x>=boundx{
		prod=1
		for i:=0; i<chunk; i++ {
			prod = prod*matrix[boundx-i][boundy+i]
		}
	}
	return prod
}

func down_right(x int, y int) uint64 {
	var prod uint64 = 0
	boundx := dim - chunk
	boundy := dim - chunk
	if y <= boundy && x<=boundx{
		prod=1
		for i:=0; i<chunk; i++ {
			prod = prod*matrix[boundx+i][boundy+i]
		}
	}
	return prod}

func prod_max(x int, y int) uint64 {
	return max(left(x,y), right(x,y), up(x,y), down(x,y), up_left(x,y), up_right(x,y), down_left(x,y), down_right(x,y))
}

func read_matrix(size int) {
	x := 0
	y := 0
	fs := bufio.NewScanner(os.Stdin)
	fs.Split(bufio.ScanWords)
	matrix = make([][]uint64, size)
	matrix[0] = make([]uint64, size)
	for fs.Scan() {
		snumber := strings.TrimLeft(fs.Text(), "0")
		fmt.Print(snumber+" ")
		n, _ := strconv.Atoi(fs.Text())
		matrix[x][y] = uint64(n)
		y = y+1
		if y == size {
			x = x+1
			y = 0
			if (x < size) {
				matrix[x] = make([]uint64, size)
			}
			fmt.Println("")
		}
	}
}

func main () {
	read_matrix(dim)

    var max_value uint64 = 0
	for i:=0; i<dim; i++ {
		for j:=0; j<dim; j++ {
			max_value = max(prod_max(i, j), max_value)
		}
	}
	fmt.Printf("Max number is %d", max_value)
}

File: euler/p11.py
#!/usr/local/bin/python3
import sys

FACTORS = 4
DIMENSION = 20
matrix = [[]]

def readMatrix(dimension):
    return [[int(n) for n in line.split()] for line in sys.stdin]

def left(y, x):
    boundx = FACTORS-1
    prod = 1
    if x>=boundx:
        for i in range(FACTORS):
            prod *= matrix[y][x-i]
    return prod

def right(y, x):
    boundx = DIMENSION-FACTORS
    prod = 1
    if x<=boundx:
        for i in range(FACTORS):
            prod *= matrix[y][x+i]
    return prod

def up(y, x):
    boundy = FACTORS-1
    prod = 1
    if y>=boundy:
        for i in range(FACTORS):
            prod *= matrix[y-i][x]
    return prod
    

def down(y, x):
    boundy = DIMENSION-FACTORS
    prod = 1
    if y<=boundy:
        for i in range(FACTORS):
            prod *= matrix[y+i][x]
    return prod

def upLeft(y, x):
    boundx = FACTORS-1
    boundy = FACTORS-1
    prod = 1
    if x>=boundx and y>=boundy:
        for i in range(FACTORS):
            prod *= matrix[y-i][x-i]
    return prod

def upRight(y, x):
    boundx = DIMENSION-FACTORS
    boundy = FACTORS-1
    prod = 1
    if x<=boundx and y>=boundy:
        for i in range(FACTORS):
            prod *= matrix[y-i][x+i]
    return prod

def downLeft(y, x):
    boundx = FACTORS-1
    boundy = DIMENSION-FACTORS
    prod = 1
    if x>=boundx and y<=boundy:
        for i in range(FACTORS):
            prod *= matrix[y+i][x-i]
    return prod

def downRight(y, x):
    boundx = DIMENSION-FACTORS
    boundy = DIMENSION-FACTORS
    prod = 1
    if x<=boundx and y<=boundy:
        for i in range(FACTORS):
            prod *= matrix[y+i][x+i]
    return prod


def prodMax(y, x):
    return max( left(y,x), \
                right(y,x), \
                up(y,x), \
                down(y,x), \
                upLeft(y,x), \
                upRight(y,x), \
                downLeft(y,x), \
                downRight(y,x))


if __name__ == "__main__":
    matrix = readMatrix(DIMENSION)
    maxValue = 0
    for i in range(DIMENSION):
        for j in range(DIMENSION):
            maxValue = max(prodMax(i,j), maxValue )
    print ("Max number is " + str(maxValue))
